### PR TITLE
Change cilium_host IPv6 address

### LINKED ISSUE #21954: Reply from pod to outside is dropped when L7 ingress policy is used
This issue is about an L7 ingress policy problem a when a pod is reached directly from outside client / via a NodePort BPF service.

Let's consider the following L7 netpol:
```
apiVersion: "cilium.io/v2"
kind: CiliumNetworkPolicy
metadata:
  name: foobar
spec:
  description: "Allow to GET on echo from outside"
  endpointSelector:
    matchLabels:
      kind: echo
  ingress:
  - fromEntities:
    - "world"
    toPorts:
    - ports:
      - port: "8080"
        protocol: TCP
      rules:
        http:
        - method: "GET"
          path: "/$"
```

When the netpol is applied, accessing the `echo` pod from outside the cluster fails with:
```
xx drop (Stale or unroutable IP) flow 0xf863e34e to endpoint 0, file bpf_host.c line 665, , identity 28143->unknown: 10.0.1.49:80 -> 192.168.34.1:32884 tcp SYN, ACK
```
The drop is triggered by https://github.com/cilium/cilium/blob/v1.12.3/bpf/bpf_host.c#L626.

What happens is that the L7 proxy sends the SYN-ACK which gets handled by `bpf_host @ cilium_host`, and then dropped. See the  `pwru` output (ifindex=9 is `cilium_host`):

```
  SKB    CPU          PROCESS                     FUNC
0xffff9a859048e800      7        [<empty>]             ip_local_out netns=4026531992 mark=0xa00 ifindex=0 proto=0 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]           __ip_local_out netns=4026531992 mark=0xa00 ifindex=0 proto=0 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             nf_hook_slow netns=4026531992 mark=0xa00 ifindex=0 proto=8 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]                ip_output netns=4026531992 mark=0xa00 ifindex=0 proto=8 mtu=0 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             nf_hook_slow netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]  apparmor_ipv4_postroute netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]         ip_finish_output netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>] __cgroup_bpf_run_filter_skb netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]       __ip_finish_output netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]        ip_finish_output2 netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=60 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]           dev_queue_xmit netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]         __dev_queue_xmit netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]             tcf_classify netns=4026531992 mark=0xa00 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]      skb_ensure_writable netns=4026531992 mark=0x0 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
0xffff9a859048e800      7        [<empty>]                kfree_skb netns=4026531992 mark=0x0 ifindex=9 proto=8 mtu=1500 len=74 10.0.1.49:80->192.168.34.1:32884(tcp)
```

The packet is sent to the `cilium_host` because of the mark and the following IP rules / routes:
```
$ ip rule list
...
10:     from all fwmark 0xa00/0xf00 lookup 2005

$ ip route show table 2005
default via 10.0.1.116 dev cilium_host
10.0.1.116 dev cilium_host scope link
```

One fix is to extend the troublesome check https://github.com/cilium/cilium/blob/v1.12.3/bpf/bpf_host.c#L626 by allowing proxy replies to `WORLD_ID`.

### LINKED ISSUE #23445: datapath: Change cilium_host IPv6 addr
Today Cilium sets the `cilium_host` IPv6 addr to the same one as a native iface. The motivation for that is listed in [the following commit message](https://github.com/cilium/cilium/pull/9062/commits/7792ffc008852765af1d553baa607ac3dc5d19fc):

> Some more context and background: for v6 (direct routing or tunnel
   mode), the cilium_host device has a globally scoped v6 address which
   matches the one of the public interface. This is due to the fact that
   we cannot add a private address with link scope like in v4 case where
   kernel can ensure that no traffic is leaked to the public:

However, recently we changed the scope of the `cilium_host` IPv4 addr from local to global - https://github.com/cilium/cilium/pull/23241, so the reasoning above _probably_ should not be used as an argument.

Having the same IPv6 addr causes some issues when the L7 proxy is involved - https://github.com/cilium/cilium/issues/21954#issuecomment-1403765405. Therefore, let's try to use the same IPAM allocation path for the IPv6 as we do today for the IPv4. The previous attempt to change that was in this non-merged PR - https://github.com/cilium/cilium/pull/16044.

### LINKED ISSUE #23461: ipsec: IPv6 reply bypasses tunnel which might lead to unintended masquerading
Hit the issue that cilium-health was reporting failing IPv6 host netns -> remote pod checks. Closer look revealed that a second reply from the remote health check pod got masqueraded by Cilium's rule:
```
[302:21728] -A CILIUM_POST_nat ! -s fd02::/120 ! -d fd02::/120 -o cilium_host -m comment --comment "cilium host->cluster masquerade" -j SNAT --to-source fc00:f853:ccd:e793::5
```
According to @pchaigno the reply should not hit the networking stack, and instead it should have redirected to the `cilium_vxlan` bypassing the rule.

To reproduce (taken from https://github.com/cilium/cilium/actions/runs/4024902232/jobs/6917456832):
```
cd cilium/
./contrib/scripts/kind.sh "" 2 "" "" "iptables" "dual"
 helm template --validate ./install/kubernetes/cilium --namespace=kube-system  --set preflight.image.useDigest=false  --set ipv4NativeRoutingCIDR=10.0.0.0/8  --set ipv6NativeRoutingCIDR=fd02::/112  --set operator.image.repository=quay.io/cilium/operator  --set encryption.enabled=true  --set ipv6.enabled=true  --set hubble.relay.image.useDigest=false  --set image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set etcd.leaseTTL=30s  --set logSystemLoad=true  --set debug.enabled=true  --set operator.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set bandwidthManager.enabled=false  --set hubble.eventBufferCapacity=65535  --set ipam.operator.clusterPoolIPv6PodCIDR=fd02::/112  --set ipv4.enabled=true  --set hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci  --set preflight.image.repository=quay.io/cilium/cilium-ci  --set hubble.listenAddress=:4244  --set bpf.masquerade=false  --set enableCnpStatusUpdates=true  --set hubble.relay.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set image.useDigest=false  --set image.repository=quay.io/cilium/cilium-ci  --set pprof.enabled=true  --set operator.image.suffix=-ci  --set bpf.preallocateMaps=false  --set sessionAffinity=false  --set hubble.enabled=true  --set operator.image.useDigest=false  --set debug.verbose=flow  --set kubeProxyReplacement=disabled  --set tunnel=vxlan  --set preflight.image.tag=28a0cd6f202b70ab81b8cabd5b5fb3cae6f63210  --set k8s.requireIPv4PodCIDR=true  --set enableCiliumEndpointSlice=true  > cilium-173e2a16d47c0211.yaml
k apply -f test/k8s/manifests/ipsec_secret.yaml -n kube-system
k apply -f cilium-173e2a16d47c0211.yaml
```
Then exec into any cilium-agent pod, and run:
```
cilium-health status -o json --probe
```

### PR DESCRIPTION
**Pull request by jschwinger233**
Originally the IPv6 of cilium_host is set to a native address. Instead, this commit uses an IPv6 allocated from IPAM for
it, which is the same way we set IPv4 address for cilium_host.

Fixes: #23445
Fixes: #21954 
Fixes: #23461

```release-note
Change cilium_host IPv6 address, use node router IPv6 instead of native node IPv6, and fixed several relative IPv6 issues.
```

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

### CODE REVIEW

**Comment by jschwinger233** [comment]
One thing left to do is kind of tricky and I may need some help. @julianwiedmann  @brb 

After simply changing the cilium_host ipv6, route table can't be updated by cilium-agent.

In a two-node cluster, I expect two crucial routes on each host:
```
$cidr_local via $ipv6_cilium_host dev cilium_host
$cidr_remote via $ipv6_ciilum_host dev cilium_host mtu 1450
```
But there are nothing.

Cilium-agent's journal log shows failure to update routes:
```
Mar 08 03:36:42 k8s1 cilium-agent[10142]: {Ifindex: 10 Dst: fd04::1:0/112 Src: fd00::b Gw: fd04::8dd1 Flags: [] Table: 0 Realm: 0}
Mar 08 03:36:42 k8s1 cilium-agent[10142]: level=warning msg="Unable to update route" error="invalid argument" interface=cilium_host local="fd00::b" nexthop="fd04::8dd1" prefix="{fd04::1:0 ffffffffffffffffffffffffffff0000}" subsys=linux-datapath
```

If I try to add route by ip(8) command line, such as `ip -6 r a fd04::0:0/112 dev cilium_host via fd04::1:d01b`, kernel will reject the request with error `Error: Gateway can not be a local address`. I believe this is the error cilium-agent hit.

I looked into the kernel source, it seems the error happens here: https://github.com/torvalds/linux/blob/v6.1-rc3/net/ipv6/route.c#L3438-L3441

I worked out a way to workaround it, it's just to temporarily delete ipv6 from cilium_host, and add ipv6 back after adding routes, something like this:
```
# temporarily delete ipv6
ip a d $ipv6 dev cilium_host
# add route for $ipv6, or kernel will complain "no route to host" when adding route for $cidr
ip r a $ipv6 dev cilium_host
# now we can add route for $cidr
ip r a $cidr_local dev cilium_host via $ipv6
ip r a $cidr_remote dev cilium_host via $ipv6 
# finally we add back $ipv6
ip a a $ipv6 dev cilium_host
```
But I don't think this approach is acceptable, as cilium-agent is supposed to maintain route table periodically, so even if we succeed in adding routes for one time, there are still loads of errors once we add back the ipv6 for cilium_host.

**Comment by brb** [comment]
@jschwinger233 :wave: A few suggestions / notes:
- Have you tried dropping the `via $GW`? 
- What is scope of the v6 addrs? global?
- The route installation is done in `pkg/datapath/linux/node.go`

**Comment by jschwinger233** [comment]
Hi @brb, thanks for the suggestions. You got up so early - -

> Have you tried dropping the via $GW?

Yes, still failed.

> What is scope of the v6 addrs? global?

global

**Comment by brb** [comment]
It's still an early morning here :sweat_smile: , but the following seems to work for the `$cidr_local via $ipv6_cilium_host dev cilium_host` case:
```
# ip -6 -o r
...
fd00:10:244:1::/64 dev cilium_host proto kernel metric 256 pref medium

# ip -6 -o a
...
3: cilium_host    inet6 fd00:10:244:1::3851/64 scope global \       valid_lft forever preferred_lft forever

# ip r g fd00:10:244:1::5629 (LOCAL POD IP)
fd00:10:244:1::5629 from :: dev cilium_host proto kernel src fd00:10:244:1::3851 metric 256 pref medium
```

**Comment by jschwinger233** [comment]
@brb  Yes I know it's 8am in the morning, and I usually don't get up at that time, sorry to let you feel that was satire.

I just tried `ip r a $ipv6 dev cilium_host` and it seems working. Shame that I didn't come up with such obvious idea, thanks for your insight, let me revise the cilium-agent.

**Comment by brb** [comment]
Please keep in mind that I haven't tried remote-cidr case

**Comment by jschwinger233** [comment]
> Please keep in mind that I haven't tried remote-cidr case

For remote cidr, we have to indicate `src $native_ipv6` to make connection between host and remote pod.

**Comment by brb** [comment]
I guess the `runtime` tests failed due to `eth0` instead of `cilium_host` v6 addr being used:
```
/home/jenkins/workspace/Cilium-PR-Runtime-net-next/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:515
Expected command: docker exec -i  app2 curl --path-as-is -s -D /dev/stderr --fail --connect-timeout 5 --max-time 20 [http://[f00d::1:1]:80/public](http://[f00d::1:1]/public) -w "time-> DNS: '%{time_namelookup}(%{remote_ip})', Connect: '%{time_connect}',Transfer '%{time_starttransfer}', total '%{time_total}'" 
To succeed, but it failed:
Exitcode: 28 
Err: Process exited with status 28
Stdout:
 	 time-> DNS: '0.000015()', Connect: '0.000000',Transfer '0.000000', total '5.001701'
Stderr:
```

This is set by https://github.com/cilium/cilium/blob/master/test/runtime/lb.go#L386

**Comment by jschwinger233** [comment]
> One thing left to do is kind of tricky and I may need some help. @julianwiedmann @brb
> 
> The problem is, after simply changing the cilium_host ipv6, route table can't be updated by cilium-agent.
> 
> In a two-node cluster, I expect two crucial routes on each host:
> 
> ```
> $cidr_local via $ipv6_cilium_host dev cilium_host
> $cidr_remote via $ipv6_ciilum_host dev cilium_host mtu 1450
> ```
> 
> But there is nothing.
> 
> Cilium-agent's journal log shows failure to update routes:
> 
> ```
> Mar 08 03:36:42 k8s1 cilium-agent[10142]: {Ifindex: 10 Dst: fd04::1:0/112 Src: fd00::b Gw: fd04::8dd1 Flags: [] Table: 0 Realm: 0}
> Mar 08 03:36:42 k8s1 cilium-agent[10142]: level=warning msg="Unable to update route" error="invalid argument" interface=cilium_host local="fd00::b" nexthop="fd04::8dd1" prefix="{fd04::1:0 ffffffffffffffffffffffffffff0000}" subsys=linux-datapath
> ```
> 
> If I try to add route by ip(8) command line, such as `ip -6 r a fd04::0:0/112 dev cilium_host via fd04::1:d01b`, kernel will reject the request with error `Error: Gateway can not be a local address`. I believe this is the same error cilium-agent hit.
> 
> I looked into the kernel source, it seems the error happens here: https://github.com/torvalds/linux/blob/v6.1-rc3/net/ipv6/route.c#L3438-L3441
> 
> I worked out a way to workaround it, it's just to temporarily delete ipv6 from cilium_host, and add ipv6 back after adding routes, something like this:
> 
> ```
> # temporarily delete ipv6
> ip a d $ipv6 dev cilium_host
> 
> # add route for $ipv6, or kernel will complain "no route to host" when adding route for $cidr
> ip r a $ipv6 dev cilium_host
> 
> # now we can add route for $cidr
> ip r a $cidr_local dev cilium_host via $ipv6
> ip r a $cidr_remote dev cilium_host via $ipv6 
> 
> # finally we add back $ipv6
> ip a a $ipv6 dev cilium_host
> ```
> 
> But I don't think this approach is acceptable, as cilium-agent is supposed to maintain route table periodically, so even if we succeed in adding routes for one time, there are still loads of errors once we add back the ipv6 for cilium_host.

@NikAleksandrov Hi Nikolay, could you shed some light on this issue? I actually realized this `via $gw` is essential to pass the connectivity test, but kernel won't allow us to do so for ipv6. Basically ipv4 has the similar route like `10.10.0.0/24 via 10.10.0.1 dev cilium_host` instead of `10.10.0.0/24 dev cilium_host`, so a more general question is, what's the key difference between those two patterns?

**Comment by jschwinger233** [comment]
> The ci-datapath failure looks legit too (https://github.com/cilium/cilium/actions/runs/4363707535/jobs/7632661159)

@brb I can reproduce this error locally, and it's seems related to `via $gw` issue. Once I add the routes with `via`, the no-policies-extra/pod-to-local-nodeport test will pass. 

Now I'm looking into how these two routes can influence L2 and L3 header in different way. Any progress will let you know.

**Comment by jschwinger233** [comment]
@brb Got some clues. 

The datapath is lxc -> stack -> cillium_host, and it's kernel stack where DNAT and SNAT both happen. DNAT replaces dest ip with real upstream pod ip, and SNAT replaces source ip with `$ipv6_cilium_host`. Compared to ipv4 datapath, this seems good, but at present cilium can't handle tcp segment with source ip `$ipv6_cilium_host`.

Let me elaborate it.

According to my observation, after a tcp syn with ipv6 header `$ipv6_cilium_host -> $ipv6_upstream_pod` is sent to cilium_host, cilium_host will reply a tcp syn with ipv6 header `$ipv6_upstream_pod -> $ipv6_native`, instead of `$ipv6_upstream_pod -> $ipv6_cilium_host`. That the place tcp connection breaks.

This is not a problem before because `$ipv6_cilium_host` == `$ipv6_upstream_pod`, but now we have to take care of it.

**Comment by NikAleksandrov** [comment]
> > One thing left to do is kind of tricky and I may need some help. @julianwiedmann @brb
> > The problem is, after simply changing the cilium_host ipv6, route table can't be updated by cilium-agent.
> > In a two-node cluster, I expect two crucial routes on each host:
> > ```
> > $cidr_local via $ipv6_cilium_host dev cilium_host
> > $cidr_remote via $ipv6_ciilum_host dev cilium_host mtu 1450
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But there is nothing.
> > Cilium-agent's journal log shows failure to update routes:
> > ```
> > Mar 08 03:36:42 k8s1 cilium-agent[10142]: {Ifindex: 10 Dst: fd04::1:0/112 Src: fd00::b Gw: fd04::8dd1 Flags: [] Table: 0 Realm: 0}
> > Mar 08 03:36:42 k8s1 cilium-agent[10142]: level=warning msg="Unable to update route" error="invalid argument" interface=cilium_host local="fd00::b" nexthop="fd04::8dd1" prefix="{fd04::1:0 ffffffffffffffffffffffffffff0000}" subsys=linux-datapath
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If I try to add route by ip(8) command line, such as `ip -6 r a fd04::0:0/112 dev cilium_host via fd04::1:d01b`, kernel will reject the request with error `Error: Gateway can not be a local address`. I believe this is the same error cilium-agent hit.
> > I looked into the kernel source, it seems the error happens here: https://github.com/torvalds/linux/blob/v6.1-rc3/net/ipv6/route.c#L3438-L3441
> > I worked out a way to workaround it, it's just to temporarily delete ipv6 from cilium_host, and add ipv6 back after adding routes, something like this:
> > ```
> > # temporarily delete ipv6
> > ip a d $ipv6 dev cilium_host
> > 
> > # add route for $ipv6, or kernel will complain "no route to host" when adding route for $cidr
> > ip r a $ipv6 dev cilium_host
> > 
> > # now we can add route for $cidr
> > ip r a $cidr_local dev cilium_host via $ipv6
> > ip r a $cidr_remote dev cilium_host via $ipv6 
> > 
> > # finally we add back $ipv6
> > ip a a $ipv6 dev cilium_host
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I don't think this approach is acceptable, as cilium-agent is supposed to maintain route table periodically, so even if we succeed in adding routes for one time, there are still loads of errors once we add back the ipv6 for cilium_host.
> 
> @NikAleksandrov Hi Nikolay, could you shed some light on this issue? I actually realized this `via $gw` is essential to pass the connectivity test, but kernel won't allow us to do so for ipv6. Basically ipv4 has the similar route like `10.10.0.0/24 via 10.10.0.1 dev cilium_host` instead of `10.10.0.0/24 dev cilium_host`, so a more general question is, what's the key difference between those two patterns?

Hi, using via $gw IMO for this case mostly influences the choice of address (as also Joe mentioned over slack). This guarantees the right address will be chosen. Although I cannot guarantee that's the reason it's used because I haven't looked into what would change and what exactly relies on it. Since you have a test that's breaking if it's not used I'd look into what changes and where the packet gets dropped.
@jschwinger233  does dropping "via $gw" break it for the IPv4 case as well?

**Review by pippolo84 — APPROVED** [review]
`/pkg/node` changes LGTM.

I think that the CI failure is related to this test [here](https://github.com/jschwinger233/cilium/blob/e571967dfe78649d3db9c8750483a83095509161/pkg/datapath/linux/node_test.go#L86) not being updated to:

`c.Assert(generatedRoute.Local, checker.DeepEquals, fakeNodeAddressing.IPv6().Router())`

after the change in `createNodeRouteSpec`

**Comment by brb** [comment]
The `k8s-1.25-kernel-4.19` failure looks legit:
> Cilium validation failed: 4m0s timeout expired: Last polled error: connectivity health is failing: cilium-agent 'cilium-pd4wj': connectivity to path 'k8s2.health-endpoint.secondary-addresses[*].http.status' is unhealthy: 'Get \"[http://[fd02::1d4]:4240/hello\":](http://[fd02::1d4]:4240/hello/) read tcp [fd02::ac]:55144->[fd02::1d4]:4240: read: connection reset by peer

Do you need help to debug it?

**Comment by jschwinger233** [comment]
> The `k8s-1.25-kernel-4.19` failure looks legit:
> 
> > Cilium validation failed: 4m0s timeout expired: Last polled error: connectivity health is failing: cilium-agent 'cilium-pd4wj': connectivity to path 'k8s2.health-endpoint.secondary-addresses[*].http.status' is unhealthy: 'Get "[http://[fd02::1d4]:4240/hello":](http://%5Bfd02::1d4%5D:4240/hello/) read tcp [fd02::ac]:55144->[fd02::1d4]:4240: read: connection reset by peer
> 
> Do you need help to debug it?

Yes lol

The failed test is `K8sDatapathServicesTest Checks E/W loadbalancing (ClusterIP, NodePort from inside cluster, etc) Tests NodePort inside cluster (kube-proxy) with IPSec and externalTrafficPolicy=Local`, and cilium status shows unhealthy, curl from k8s1 host to health@k8s2 gets tcp RST `connection reset by peer`

I tcpdumped trafifc and found:
1. tcp 3-way handshake is completed and connection is established
2. client sends http request payload succesfully, and health server also receives http request and send back the response
3. but client never gets http response
4. server will receive several tcp retransmission of http request, followed by a tcp reset from client
5. client will re-send the tcp request, before it sees a tcp reset from server

Both ends see tcp reset from other side, and the http response can't reach client, in the meantime there's no `drop` log in `cilium monitor`. 

It'd be helpful if you can throw some light.

**Comment by brb** [comment]
> but client never gets http response

Maybe you could check with `pwru` what happens with the response. There have been quite a few IPv6 related issue recently. E.g., https://github.com/cilium/cilium/issues/2348, https://github.com/cilium/cilium/issues/23461. Both have some examples how to debug with `pwru`. I'd suggest to reproduce on Kind, as then you can inspect all nodes packets from your host.

**Comment by brb** [comment]
Actually https://github.com/cilium/cilium/issues/23461 symptoms are very similar to yours.

**Comment by julianwiedmann** [comment]
The ipsec changes should be fixing https://github.com/cilium/cilium/issues/23481, right? Could we add a revert of https://github.com/cilium/cilium/commit/b1b9fbd5ea46520289a1c93425eed45d6308d75a and see if that scenario works as expected now?

(imho it would be best to do that in a separate PR)

**Comment by jschwinger233** [comment]
> The ipsec changes should be fixing #23481, right? Could we add a revert of [b1b9fbd](https://github.com/cilium/cilium/commit/b1b9fbd5ea46520289a1c93425eed45d6308d75a) and see if that scenario works as expected now?
> 
> (imho it would be best to do that in a separate PR)

Of course, I'll get right on it.

**Review by brb — APPROVED** [review]
:exploding_head: amazing!

**Review by NikAleksandrov — APPROVED** [review]

File: pkg/datapath/iptables/iptables.go [hunk 1]
```diff
@@ -1611,31 +1611,36 @@ func (m *IptablesManager) addCiliumAcceptXfrmRules() error {
...
-		return err
-	}
-	if err := insertAcceptXfrm("nat", ciliumPreNatChain); err != nil {
-		return err
-	}
-	if err := insertAcceptXfrm("nat", ciliumOutputNatChain); err != nil {
-		return err
+
+	for _, insertOpt := range insertOpts {
+		if err := insertAcceptXfrm(ip4tables, insertOpt.table, insertOpt.chain); err != nil {
```
Comment (pchaigno) [inline pkg/datapath/iptables/iptables.go]: Should we skip this if IPv4 is disabled?

**Review by pchaigno — COMMENTED** [review]
Nice work!

If we are fixing IPsec+IPv6, shouldn't we be able to reenable some tests to cover it?

**Review by pchaigno — APPROVED** [review]
> If we are fixing IPsec+IPv6, shouldn't we be able to reenable some tests to cover it?

That can probably be a follow up PR if we don't want to delay merging this one.

**Comment by jschwinger233** [comment]
> That can probably be a follow up PR if we don't want to delay merging this one.

That's right, I'm going to do that.

I had a discussion with @julianwiedmann and we decided to split this PR from reverting 3ed62d51d5ac27bbe2c60372850d4ec78c17bb6e, in case there is any hidden issue that may forces us to revert this PR.

I opened a draft PR #24807 to check if this PR has fixed the 2005 route table issues.

(Also Julian suggested re-organizing the commit messages with more details and impact descriptions; I think it's a good idea and I did it.

**Review by pchaigno — APPROVED** [review]
:heart: for the improved commit descriptions.

**Comment by julianwiedmann** [comment]
Just as note for the last commit (https://github.com/cilium/cilium/pull/24208/commits/edfd1287db9049ecb25d3ab906bf3ac449bb0f1c). No need to fix unless you need to push anyways.

>Within a cluster without KPR disabled, an ipv6 traffic from host1 to
nodeport-svc@host2 is expected to go through following steps on host2:

This should probably be `Within a cluster **with** KPR disabled` - as we're fixing a scenario with nodeport service translation by kube-proxy (not by Cilium's kube-proxy replacement).

**Comment by julianwiedmann** [comment]
> > That can probably be a follow up PR if we don't want to delay merging this one.
> 
> That's right, I'm going to do that.
> 
> I had a discussion with @julianwiedmann and we decided to split this PR from reverting [3ed62d5](https://github.com/cilium/cilium/commit/3ed62d51d5ac27bbe2c60372850d4ec78c17bb6e), in case there is any hidden issue that may forces us to revert this PR.
> 
> I opened a draft PR #24807 to check if this PR has fixed the 2005 route table issues.
> 
> (Also Julian suggested re-organizing the commit messages with more details and impact descriptions; I think it's a good idea and I did it.

great stuff, thank you! :+1:

**Comment by jschwinger233** [comment]
I think the both CI failures are flakes.

For [k8s-1.16-kernel-4.19 test](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.16-kernel-4.19/877/), I ran `K8S_NODE=2 K8S_VERSION=1.16 KERNEL=419 NFS=1 ginkgo --focus="K8sDatapathConfig MonitorAggregation Checks"    --tags=integration_tests -- -cilium.provision=true -cilium.holdEnvironment=true -cilium.passCLIEnvironment=true` locally, and it passed test.
For [runtime test](https://jenkins.cilium.io/job/Cilium-PR-Runtime-net-next/6000/), I ran `sudo make tests-privileged` locally and also it looks good.

**Comment by pchaigno** [comment]
> I think the both CI failures are flakes.

If you can point to the opened issues for those flakes, then we can mark ready to merge.

**Comment by jschwinger233** [comment]
> If you can point to the opened issues for those flakes, then we can mark ready to merge.

runtime failure hits this one https://github.com/cilium/cilium/issues/24326

But I couldn't find an issue for k8s-1.16-kernel-4.19 error `monitor aggregation did not result in correct number of TCP notifications`. Just ran this e2e test locally again, and I still can't reproduce it.

**Comment by pchaigno** [comment]
> But I couldn't find an issue for k8s-1.16-kernel-4.19 error `monitor aggregation did not result in correct number of TCP notifications`. Just ran this e2e test locally again, and I still can't reproduce it.

It's happening on master as well (found via [the CI dashboard](https://datastudio.google.com/u/1/reporting/1kYVnPCU-tJCRUV2E3Pf044NFNnlFZgmW/page/IcyMB)):
- https://jenkins.cilium.io/job/cilium-master-k8s-1.16-kernel-4.19/4451/testReport/junit/Suite-k8s-1/16/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/
- https://jenkins.cilium.io/job/cilium-master-k8s-1.16-kernel-4.19/4445/testReport/junit/Suite-k8s-1/16/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/
- https://jenkins.cilium.io/job/cilium-master-k8s-1.16-kernel-4.19/4420/testReport/junit/Suite-k8s-1/16/K8sDatapathConfig_MonitorAggregation_Checks_that_monitor_aggregation_flags_send_notifications/

Could you create an issue for it, with an example artifact from one of the failing tests?

**Comment by jschwinger233** [comment]
> Could you create an issue for it, with an example artifact from one of the failing tests?

Here it is: https://github.com/cilium/cilium/issues/24840

Thank you for help :+1:

File: bpf/Makefile [hunk 2]
```diff
@@ -58,7 +58,7 @@ LB_OPTIONS = \
 	-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE: \
 	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
-	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC:-DENABLE_ENCAP_HOST_REMAP: \
+	-DENABLE_IPV4:-DENABLE_IPV6:-DENCAP_IFINDEX:-DTUNNEL_MODE:-DENABLE_IPSEC: \
```
Comment (julianwiedmann) [inline bpf/Makefile:61]: Once merged, could you open a follow-on PR to remove this duplication? No need to fix it here unless you need to push for some other reason :+1:.

**Review by julianwiedmann — APPROVED** [review]

File: bpf/Makefile (hunk 2 above)
Comment (jschwinger233) [reply bpf/Makefile:61]: For sure! I took 30 seconds to understand what duplication you were saying LOL

**Comment by dylandreimerink** [comment]
/test-1.16-4.19

**Pull request merged**